### DISCUSSION TITLE
Quasar CLI + webpack v3.4.5 + Pinia outside component not working

### DISCUSSION THREAD

**Discussion by miguelrrsantos** [quasarframework/quasar#12927] [General - Components / Directives / etc] [answered]
Hello.

I created a new project through Quasar CLI and selected pinia as my state store. 

So far, i have a stores folder with an index.js with the following content:
_src/stores/index.js_
```javascript
import { store } from 'quasar/wrappers'
import { createPinia } from 'pinia'

/*
 * If not building with SSR mode, you can
 * directly export the Store instantiation;
 *
 * The function below can be async too; either use
 * async/await or return a Promise which resolves
 * with the Store instance.
 */

export default store((/* { ssrContext } */) => {
  const pinia = createPinia()

  // You can add Pinia plugins here
  // pinia.use(SomePiniaPlugin)

  return pinia
})
```

Also, created the session store:
_src/stores/session/index.js_
```javascript
import { defineStore } from 'pinia'

export const useSessionStore = defineStore('session', {
    state: () => ({
        user: null,
        userConfiguration: null,
    }),

    getters: {
        getUserConfiguration: (state) => state.userConfiguration,
        isAuthenticated: (state) => !!state.user,
    },

    actions: {
        setUser(user) {
            this.user = user
        },
        setUserConfiguration(userConfiguration) {
            this.userConfiguration = userConfiguration
        },
    },
})
```

I made a Login component that calls a javascript authentication service:
_src/pages/Login.vue_ 
```javascript

import autenticationService from 'src/services/authentication/index'
...
let response = await autenticationService.login(parameters)
...

```

The authentications service will try to authenticate:
_src/services/authentication/index.js_
```javascript
import session from './session'

export default {

    ...

    async login(credentials) {
        let response = await this.authenticate(credentials)

        if (response.data.status === 'success') {
            await session.saveSession(response.data.data)
        }

        return response
    },
}
```
When the session service is called, it tries to use the session store:
_src/services/authentication/session.js_
```javascript
import { useSessionStore } from 'src/stores/session'
export default {
    async setConfigurations(user) {
        const sessionStore = useSessionStore()
        console.log(sessionStore)
        sessionStore.setUser = user
    },

    async setUserConfiguration(userConfiguration) {
        const sessionStore = useSessionStore()
        sessionStore.setUserConfiguration = userConfiguration
    },

    async saveSession(data) {
        localStorage.setItem('token', data.jwtToken)

        await this.setConfigurations({
            first_name: 'User',
            last_name: 'Name',
        })
        await this.setUserConfiguration({
            locale: 'pt-PT',
        })
    }
}
```

I do not get any error and my console.log shows the sessionStore object. However, the pinia state is not being updated:
![image](https://user-images.githubusercontent.com/22508733/160036233-2d9d69f9-f4b5-44ff-b491-17fb54f09910.png)

What am i doing wrong?

I already made another experiment in Login.vue component:
```javascript
import { useSessionStore } from 'src/stores/session'
import autenticationService from 'src/services/authentication/index'
...

                let response = await autenticationService.login(parameters)

                sessionStore.setUser({
                    first_name: 'User',
                    last_name: 'Name',
                })
                sessionStore.setUserConfiguration({
                    locale: 'pt-PT',
                })
```

And the result is that the state is updated:
![image](https://user-images.githubusercontent.com/22508733/160036599-1a1bddb6-3fa7-45f2-bdd9-55eabe7eb63a.png)

But for this purpose i want to use the javascript service files instead of the component. Anyone can help me?

Best regards,
Miguel.

**Comment by TobyMosque** [comment, marked as answer]
read that, u'll understand your getter and actions are umecessary, all what u need is your state.

https://pinia.vuejs.org/cookbook/migration-vuex.html

Regarding access the store outside of a component, u need to pass the Pinia instancia to your store function.

https://pinia.vuejs.org/cookbook/migration-vuex.html

IMHO your service as it is is umecessary too, u can either transform that in a composable or move the methods to the actions (my recommendations).

https://vuejs.org/guide/reusability/composables.html

**Reply by miguelrrsantos** [reply]
Hello.

Thanks for your help @TobyMosque .

I still have the following doubts:

1. Should i import the session store to _src/stores/index.js_ ?
2. In the documentation present in [https://pinia.vuejs.org/cookbook/migration-vuex.html#usage-outside-components](https://pinia.vuejs.org/cookbook/migration-vuex.html#usage-outside-components) i didn't get any example how to pass pinia instance to my store function. Can you give me an example?
3. What are the benefits of not using getters and actions?

Best regards,
Miguel.

**Reply by TobyMosque** [reply]
> Should i import the session store to src/stores/index.js ?

as i said, u can either transform your service in a composable or merge them with the store.:

**src/stores/session.js**
```js
import { defineStore } from 'pinia'

export const useSessionStore = defineStore('session', {
  state: () => ({
    user: null,
    userConfiguration: null,
    token: ''
  }),
  // https://github.com/prazdevs/pinia-plugin-persistedstate
  persist: {
    key: 'session-store',
    storage: window.sessionStorage,
    paths: ['token'],
  }
})
```
**src/composables/session.js**
```js
import { useSessionStore } from 'stores/session'

export function useSession({ pinia } = {}) {
  const sessionStore = useSessionStore(pinia);
  const { token, user, userConfiguration } = storeToRefs(sessionStore);
  async function saveSession(data) {
    token.value = data.jwtToken    
    user.value = {
      first_name: 'User',
      last_name: 'Name',
    }
    userConfiguration.value = {
      locale: 'pt-PT',
    }
  }
  return {
    saveSession
  }
}
````
**src/composables/auth.js**
```js
import { useSession } from './session'

// this composable can be a store defineStore('auth', {})
export function useAuth({ pinia } = {}) {
  const { saveSession } = useSession({ pinia });
  async function authenticate(credentials) {
    return doSomething();
  }
  async function login(credentials) {
    let response = await authenticate(credentials)
    // i hope u API isn't retuning 200 when that would return a 422, 500 or 503
    if (response.data.status === 'success') {
      await saveSession(response.data.data)
    }
    return response
  }
  return {
    authenticate,
    useSessionStore
  }
}
````

as u can see, your compose your reusable code, so there is no need of global services.
and doing that, u don't need to expose the store, router, i18n globally (what had some drawbacks).

u can call this method from the setup (script):
```html
<template>
  <q-btn :label="$t('login')" @click="login" />
</template>
<script setup>
import { useAuth } from 'src/composables/auth'
const { login } = useAuth()
return {
  login
}
</script>
```

> In the documentation present in https://pinia.vuejs.org/cookbook/migration-vuex.html#usage-outside-components i didn't get any example how to pass pinia instance to my store function. Can you give me an example?

wrong link:
https://pinia.vuejs.org/core-concepts/outside-component-usage.html

> What are the benefits of not using getters and actions?

Getters are useful when u wanna to transform the state to something else, u use that in the same way as u use the computed.

actions are useful when u need to do a task who requires more of one line of code, like query a database and update your state with the response. any way, with Pinia actions can be or not be async, so u can do asynchronous and synchronous tasks.

for example, the who login/saveSession is a good candidate to be transformed in a action:

```js
import { defineStore } from 'pinia'
import jwtDecode from 'jwt-decode'

export const useSessionStore = defineStore('session', {
  state: () => ({
    user: null,
    userConfiguration: null,
    token: ''
  }),
  getters: {
    decoded (state) {
      if (!state.token) {
        return undefined
      }
      return jwtDecode(state.token)
    },
    expireAt () {
      if (!this.decoded || !this.decoded.exp) {
        return undefined
      }
      const expiration = this.decoded.exp * 1000
      return new Date(expiration)
    }
  },
  actions: {
    // a asynchronous task
    async login (credentials) {
      const { data } = this.$api.post('/login', credentials);
      // i hope u API isn't retuning 200 when that would return a 422, 500 or 503
      if (response.data.status !== 'success') {
        return;
      }
      this.value = data.jwtToken    
      this.value = {
        first_name: 'User',
        last_name: 'Name',
      }
      this.userConfiguration = {
        locale: 'pt-PT',
      }
    },
    // a synchronous task
    isLogged () {
      const now = new Date()
      return this.expireAt && this.expireAt > now
    },
  },
  // https://github.com/prazdevs/pinia-plugin-persistedstate
  persist: {
    key: 'session-store',
    storage: window.sessionStorage,
    paths: ['token'],
  }
})
```

**Reply by miguelrrsantos** [reply]
@TobyMosque thank you very much for the effort in how the answer was clear when presenting these examples. I'm migrating from vue 2 to vue 3 and i'm still catching the way composition api and other stuff works. 

I was trying to implement the same logic that i had in previous solutions but indeed i need to change the way things can be implemented now.

Best regards,
Miguel Real.
